File: implementations/go/examples/internal/domain/card/card.go
package card

import (
	"time"

	"github.com/example/account-service/internal/common"
)

// Card is the Aggregate Root representing an issued card. Card has no way
// to know whether the linked account (AccountID) is active — whether
// issuance is allowed (account status) is determined by the Application
// layer via a synchronous lookup through AccountAdapter (ACL) before it
// calls the IssueCard factory (see cross-domain.md).
type Card struct {
	CardID    string
	AccountID string
	OwnerID   string
	Brand     string
	Status    Status
	CreatedAt time.Time
	// LastStatementSentMonth is the period for which the monthly card usage
	// statement was last sent (in "2006-01" format, e.g. "2026-07"). An
	// empty string means one has never been sent. Same idiom as
	// Account.LastInterestPaidAt — this single field is enough to determine
	// "has this period already been sent" (Level 1 — inherent idempotency),
	// so if the same period's batch Task is re-run at-least-once, it becomes
	// a natural no-op.
	LastStatementSentMonth string
}

// IssueCard issues a new card in ACTIVE status (corresponds to nestjs's Card.issue()).
func IssueCard(accountID, ownerID, brand string) *Card {
	return &Card{
		CardID:    common.NewID(),
		AccountID: accountID,
		OwnerID:   ownerID,
		Brand:     brand,
		Status:    StatusActive,
		CreatedAt: time.Now(),
	}
}

// Reconstitute restores a row read from storage into a domain object (restored as-is, without invariant checks).
func Reconstitute(cardID, accountID, ownerID, brand string, status Status, createdAt time.Time, lastStatementSentMonth string) *Card {
	return &Card{
		CardID:                 cardID,
		AccountID:              accountID,
		OwnerID:                ownerID,
		Brand:                  brand,
		Status:                 status,
		CreatedAt:              createdAt,
		LastStatementSentMonth: lastStatementSentMonth,
	}
}

// Suspend suspends a card. An already-cancelled card cannot be suspended,
// and re-suspending an already-suspended card is also meaningless, so both
// are errors (same rule as nestjs's Card.suspend()).
func (c *Card) Suspend() error {
	if c.Status == StatusCancelled {
		return ErrCancelledCardCannotBeSuspended
	}
	if c.Status == StatusSuspended {
		return ErrAlreadySuspended
	}
	c.Status = StatusSuspended
	return nil
}

// Cancel cancels a card. Re-cancelling an already-cancelled card is an
// error. It can be cancelled from either ACTIVE or SUSPENDED status (same as
// nestjs's Card.cancel()).
func (c *Card) Cancel() error {
	if c.Status == StatusCancelled {
		return ErrAlreadyCancelled
	}
	c.Status = StatusCancelled
	return nil
}

// MarkStatementSent records that the monthly usage statement for period
// (in "2006-01" format) has been sent. If it's already recorded for the same
// period, nothing changes and it returns false (Level 1 — inherent
// idempotent no-op) — the caller (SendCardUsageStatementHandler) uses this
// signal to decide whether a Save is needed, based on whether this was
// actually a new send. Sending the notification itself (the SES call) is an
// external side effect Card knows nothing about, so the caller must call
// this method only after the send succeeds — that ordering ensures a retry
// (at-least-once) after a failed send actually attempts to send again.
func (c *Card) MarkStatementSent(period string) bool {
	if c.LastStatementSentMonth == period {
		return false
	}
	c.LastStatementSentMonth = period
	return true
}

File: implementations/fastapi/examples/src/card/domain/card.py
from __future__ import annotations

from datetime import datetime
from typing import Union

from ...common.generate_id import generate_id
from .card_status import CardStatus
from .errors import (
    CancelledCardCannotBeSuspendedError,
    CardAlreadyCancelledError,
    CardAlreadySuspendedError,
)
from .events import CardStatementSent

CardDomainEvent = Union[CardStatementSent]


class Card:
    def __init__(
        self,
        card_id: str,
        account_id: str,
        owner_id: str,
        brand: str,
        status: CardStatus,
        created_at: datetime,
        last_statement_sent_month: str | None = None,
    ) -> None:
        self.card_id = card_id
        self.account_id = account_id
        self.owner_id = owner_id
        self.brand = brand
        self.status = status
        self.created_at = created_at
        # The Level 1 idempotency marker for the monthly card-statement delivery batch
        # ("YYYY-MM") — the same design as Account.last_interest_paid_at (see "Event Handler
        # Idempotency" in domain-events.md).
        self.last_statement_sent_month = last_statement_sent_month
        self._events: list[CardDomainEvent] = []

    @classmethod
    def issue(cls, account_id: str, owner_id: str, brand: str) -> Card:
        # The Card Aggregate cannot know whether the linked account is active — whether
        # issuance is allowed (account status) is decided by the Application layer via a
        # synchronous lookup through AccountAdapter (ACL), before it calls this factory.
        return cls(
            card_id=generate_id(),
            account_id=account_id,
            owner_id=owner_id,
            brand=brand,
            status=CardStatus.ACTIVE,
            created_at=datetime.utcnow(),
        )

    def suspend(self) -> None:
        if self.status == CardStatus.CANCELLED:
            raise CancelledCardCannotBeSuspendedError()
        if self.status == CardStatus.SUSPENDED:
            raise CardAlreadySuspendedError()
        self.status = CardStatus.SUSPENDED

    def cancel(self) -> None:
        if self.status == CardStatus.CANCELLED:
            raise CardAlreadyCancelledError()
        self.status = CardStatus.CANCELLED

    def send_statement(self, period: str, payment_count: int, total_amount: int, email: str) -> None:
        """An Aggregate method invoked system-driven by the monthly card-statement delivery
        batch (Task Queue → card.statement.send). It's called after the Application layer
        has already finished computing the payment count/total via a synchronous lookup
        through PaymentAdapter (ACL) — Card itself doesn't know the Payment BC.

        Idempotency is guaranteed by a single field, `last_statement_sent_month` (has it
        already been sent this month) (Level 1) — a complete no-op if already processed
        this month.
        """
        if self.last_statement_sent_month == period:
            return
        self.last_statement_sent_month = period
        self._events.append(
            CardStatementSent(
                card_id=self.card_id,
                account_id=self.account_id,
                owner_id=self.owner_id,
                email=email,
                period=period,
                payment_count=payment_count,
                total_amount=total_amount,
                sent_at=datetime.utcnow(),
            )
        )

    def pull_events(self) -> list[CardDomainEvent]:
        events, self._events = self._events, []
        return events
